tensorrt_llm/_torch/pyexecutor/py_executor.py

```diff
@@ -1469,6 +1510,236 @@ def _executor_loop_overlap(self):
+
+                    logger.debug(
+                        f'has {len(self.active_requests)} active_request, '
+                        f'scheduled {len(scheduled_batch.context_requests)} context requests'
+                    )
+                    if scheduled_batch.batch_size == 0 \
+                            and (len(ctx_requests) > 0 or self.executor_request_queue.get_waiting_queue_size() > 0):
+                        self.gen_request_cv.wait()
+                        continue
+
+                finished_requests = []
+
+                if scheduled_batch.batch_size > 0 or (
+                        self.enable_attention_dp and self.dist.tp_size > 1):
+                    with self.sm_disagg_lock:
+                        self.resource_manager.prepare_resources(scheduled_batch)
+
+                    with torch.cuda.stream(stream):
+                        batch_outputs = self._forward_step(
+                            scheduled_batch, model_engine=self.ctx_model_engine)
+                        sample_state = self._sample_async(
+                            scheduled_batch, batch_outputs)
+                        # To avoid long sync time in critical section below
+                        sample_state.sampler_event.synchronize()
+
+                    with self.sm_disagg_lock:
+                        self._update_request_states(scheduled_batch)
+                        self._update_requests(sample_state,
+                                              self.resource_manager)
+                        self._handle_canceled_requests()
+                        finished_requests = self._handle_responses()
+
+                        attn_metadata = getattr(self.ctx_model_engine,
+                                                'attn_metadata', None)
+                        kv_cache_dtype_byte_size = getattr(
+                            self.ctx_model_engine, 'kv_cache_dtype_byte_size',
+                            None)
+                        self.resource_manager.update_resources(
+                            scheduled_batch, attn_metadata,
+                            kv_cache_dtype_byte_size)
+                        if self.enable_kv_cache_events:
+                            self._add_kv_cache_events()
+
+                        self.ctx_request_cv.notify()
+
+                if self.enable_iter_perf_stats and sample_state is not None:
+                    iter_stats.iter_counter = self.ctx_model_engine.iter_counter
+                    iter_stats.inflight_batching_stats.num_ctx_tokens = self.ctx_model_engine.iter_states[
+                        'num_ctx_tokens']
+                    self._process_iter_stats(
+                        finished_requests, self.active_requests,
+                        BatchState(sample_state=sample_state,
+                                   iter_stats=iter_stats,
+                                   iter_start_time=iter_start_time))
+
+    def _executor_loop_sm_disagg_gen_overlap(self, stream):
```

Comment:
There is a large amount of duplicated code from the main executor loop with overlap. Could we refactor to avoid so much duplication? This duplicated code makes it very hard to maintain.